stdlib/Dates/src/periods.jl

```diff
@@ -170,7 +131,7 @@ struct CompoundPeriod <: AbstractTime
     function CompoundPeriod(p::Vector{Period})
         n = length(p)
         if n > 1
-            sort!(p, rev=true, lt=periodisless)
+            sort!(p, rev = true, by = days ∘ oneunit)
```

Comment:
Hmmm, what's going on here? Can you explain the magic?

Reply:
Sure!

We sort periods in decreasing order (`rev = true`) according to the length of the _period_ (e.g `Week(1)` before `Hour(240)`). There are many ways to get the length of a period `p` and `days(oneunit(p))` seems like a reasonable choice. How many days is one `Second`/`Week`/`Year`? This is sometimes a `Flaot64` and sometimes an `Int64` but we already don't have type stability because we are working with `Vector{Period}` and this constant folds once the type is known.

Reply:
That explanation feels a bit long to include as a comment, but we could if you think it is appropriate.

Reply:
Yeah adding a comment on how this works is probably a good idea, specially given that the code it replaces is pretty obvious in how it works

Reply:
Does this explain it adequitely?
```
# We sort periods in decreasing order (rev = true) according to the length of
# the period's type (by = days ∘ oneunit). We sort by type, not value, so that
# we can merge equal types.
#
# This works by computing how many days are in a single period, and then sorting
# by that. For example, (days ∘ oneunit)(Week(10)) = days(oneunit(Week(10))) =
# days(Week(1)) = 7, which is less than (days ∘ oneunit)(Month(-2)) = 30.436875
```

Reply:
Even if this is less obvious, I think its a worthwhile change because it takes the number of new methods to be implemented for a new `Period` type to support `CompundPeriod`s down from O(n) to 0.

Reply:
That sounds quite clear 👍

Reply:
Thanks!